### PR TITLE
Allow Quantize KV Cache over Multiple Dimensions

### PR DESCRIPTION
**Pull request by morgandu**
# Checklist
This PR allows quantize kv cache over specified axis or axes
- [x] Microbenchmark - Performance
- [x] E2E Serving - Accuracy and Performance

# Setup

- v5e8
- llama2-7b
- quant_mode
    - for w-i8-kv-i8
- model_mode
    - base for performance
    - chat for accuracy

# Results and Analysis

### Annotation
- b: batch
- t: query_length
- h: query_heads
- d: kv_dimension
- s: kv_length
- k: kv_heads

### Layout
- 0123 | bthd | bskd
- 0213 | bhtd | bksd
- 0231 | bhdt | bkds
- 1203 | thbd | skbd

### KV Cache Quantization Axis/Axes
- d: existing quantization over kv_dimenison
- hd: new quantization over kv_heads and kv_dimension

## Summary

- Expected Acc drop, about 1.21% due to quantization axis change, from 42.4362 to 41.9234
- Perf improved 2x plus from 2488.511837 tokens/s to 5829.69 tokens/s due to composite factors including:
  - compute layout change, PR #709
  - cache layout tuning, PR #667
  - reshape_q, PR #690
  - kv cache quantization change, this PR

#### Exiting quantization over kv_dimenison, i.e. --kv_quant_axis=d

```
{'rouge1': 42.4362, 'rouge2': 20.1052, 'rougeL': 27.2529, 'rougeLsum': 39.9515, 'gen_len': 1133702, 'gen_num': 995}
```

#### New quantization over kv_heads and kv_dimenison, i.e. --kv_quant_axis=hd

```
{'rouge1': 41.9234, 'rouge2': 19.601, 'rougeL': 26.6612, 'rougeLsum': 39.5181, 'gen_len': 1156496, 'gen_num': 995}
```

###### Cache layout 0213-0213

With default cache layout, i.e. same layout as compute

```
Successful requests: 995
Benchmark duration: 179.016534 s
Total input tokens: 217011
Total generated tokens: 979107
Request throughput: 5.56 requests/s
Input token throughput: 1212.24 tokens/s
Output token throughput: 5469.37 tokens/s
```
xprof: https://xprof.corp.google.com/overview_page/morgandu-15535915910542734436

###### Cache layout 0231-0213

The cache layout tuning gave us top cache layout 0231-0213, where the default cache layout 0213-0213 was ranked the 2nd by ar throughut.

```
Successful requests: 995
Benchmark duration: 167.951811 s
Total input tokens: 217011
Total generated tokens: 979107
Request throughput: 5.92 requests/s
Input token throughput: 1292.10 tokens/s
Output token throughput: 5829.69 tokens/s
```

xprof: https://xprof.corp.google.com/overview_page/morgandu-8809708365303398508

###### Cache layout 1203-1203
With old compute layout 0123, old cache layout 1203-1203, and existing kv cache quantization, serving throughput was at 2488.511837 tokens/s.

Cache layout tuning gave us better top performant cache layout, though all of them were struggling between self-attention and cache update.
###### Cache layout 0231-0231
top layout, fast self-attention with slow cache update, yielding serving throughput 3333.375208 tokens/s, xprof: https://xprof.corp.google.com/overview_page/morgandu-17352860818423398257

###### Cache layout 0231-2130
top layout, slow self-attention with fast cache update, yielding serving throughput 3037.178915 tokens/s, xprof: https://xprof.corp.google.com/overview_page/morgandu-2974435458070151506

###### Reshaping Q with cache layout 0231-0213
Reshaping Q during ar forced compiler not broadcasting query to full sequence length and pushed compute to MXU, this bumped serving throughput to 3598.09 tokens/s, however, still suffer from long update with the optimal layout 0231-0213. 

###### Cache layout 0213-0213
New compute layout with default cache layout 0213-0213 yielding serving throughput 2112.00 tokens/s, xprof: https://xprof.corp.google.com/overview_page/morgandu-1040975859287826924

###### Cache layout 0231-0213
Layout tuning give us best cache layout 0231-0213, yielding 2985.70 tokens/s, xprof: https://xprof.corp.google.com/overview_page/morgandu-12972597741110718440

### CODE REVIEW

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -61,6 +61,7 @@ save_config_to_gcs: False
 dtype: "bfloat16"
 quantization: "" #defaults to no quantization, i.e. bf16. possible alternative setting is 'int8' or use fp8 to run with 8-bit floating-point GeMMs on NVIDIA GPUs.
 quantize_kvcache: False
+kv_quant_axis: "kd"
```
Comment (vipannalla) [inline MaxText/configs/base.yml:64]: Can you add comment with possible values for this param?

File: MaxText/configs/base.yml [hunk 2]
```diff
@@ -317,10 +318,14 @@ inference_microbenchmark_log_file_path: ""
...
-ar_key_axis_order: "1,2,0,3"
-ar_value_axis_order: "1,2,0,3"
+prefill_cache_axis_order: "1,2,0,3"
+ar_cache_axis_order: "1,2,0,3"
+
+# Compute layout control
+# Default layout: 0,1,2,3 ; BATCH, LENGTH, HEAD, D_KV
+# Currently only support compute layout: 0,1,2,3 and 0,2,1,3
+compute_axis_order: "0,1,2,3"
+
```
Comment (vipannalla) [inline MaxText/configs/base.yml:328]: I thought these changes were already reviewed and merged?

**Review by vipannalla — COMMENTED** [review]
Looks good, but there is lot of code duplicated from the compute_axis PR. Can you rebase from latest?

**Comment by morgandu** [comment]
> Looks good, but there is lot of code duplicated from the compute_axis PR. Can you rebase from latest?

It was not ready for review, PR #667 just got merged.

Rebased to main and ready for review now.

File: MaxText/configs/base.yml (hunk 1 above)
Comment (vipannalla) [reply MaxText/configs/base.yml:70]: @singh-mitali, FYI

**Review by vipannalla — APPROVED** [review]
LGTM

File: MaxText/layers/attentions.py [hunk 3]
```diff
@@ -411,97 +416,38 @@ def wv_product(self, attn_weights: Array, value: Array, model_mode: str) -> Arra
...
-
-    Annotations:
-      b: batch size
-      s: key / value length
-      n_kv: number of kv heads, sometimes annotated as k
-      d: head / kv dimension
-
-    """
-    return tuple([kv_shape[i] for i in cached_axis_order])
+  def swap(self, items: tuple[Any, Any, Any, Any], axis_order: AxisIdxes) -> tuple[Any, Any, Any, Any]:
```
Comment (gobbleturk) [inline MaxText/layers/attentions.py:425]: Can we rename this method permute_tuple or transpose_tuple?

File: MaxText/layers/attentions.py [hunk 4]
```diff
@@ -411,97 +416,38 @@ def wv_product(self, attn_weights: Array, value: Array, model_mode: str) -> Arra
...
+    return tuple([items[i] for i in axis_order])
 
-  def _get_prefill_cache(self, batch, heads, kv_head_size, quantize_kvcache):
-    dtype = jnp.int8 if quantize_kvcache else jnp.bfloat16
+  def _get_prefill_cache_vars(self, batch, heads, kv_head_size):
 
+    dtype = jnp.int8 if self.quantize_kvcache else jnp.bfloat16
```
Comment (gobbleturk) [inline MaxText/layers/attentions.py:430]: I guess this isn't changed in this PR, but should this be
```
jnp.int8 if self.quantize_kvcache else self.dtype
```

File: MaxText/configs/base.yml (hunk 1 above)
Comment (gobbleturk) [reply MaxText/configs/base.yml:70]: Can we use more descriptive names here? I realize we don't have a great name for head dim - maybe dkv - so the three could be "dkv", "heads_and_dkv", "dkv_broadcast"

I don't see the differnce between "dkv" and "dkv_broadcast" in behavior/code though? What is meant by broadcast here?

Comment (gobbleturk) [reply MaxText/configs/base.yml:70]: I had a hard time parsing the first sentence of this comment, I think what is meant is that kv_quant_axis is only applied when quantize_kvcache is True?

Comment (gobbleturk) [reply MaxText/configs/base.yml:70]: Maybe also add a sentence or two explaining the tradeoff - correct/modify this if wrong:

More accurate quantiziation is achieved by using finer grained scale calibrations across more axes, however this comes with the cost of additional compute time.

**Review by gobbleturk — CHANGES_REQUESTED** [review]

File: MaxText/configs/base.yml [hunk 5]
```diff
@@ -61,6 +61,13 @@ save_config_to_gcs: False
 dtype: "bfloat16"
 quantization: "" #defaults to no quantization, i.e. bf16. possible alternative setting is 'int8' or use fp8 to run with 8-bit floating-point GeMMs on NVIDIA GPUs.
 quantize_kvcache: False
+# Valid kv_quant_axis values:
+#   - "" is valid only when quantize_kvcache is False
+#   - "dkv" over kv_dimensions
+#   - "heads_and_dkv" over kv_heads and kv_dimensions
+# Default to "heads_and_dkv" for faster compution, any valid kv_quant_axis non "" value will not be used when quantize_kvcache is False
```
Comment (gobbleturk) [inline MaxText/configs/base.yml:68]: Nit: Reword 

any valid kv_quant_axis non "" value will not be used when quantize_kvcache is False

to

kv_quant_axis is not used when quantize_kvcache is False

**Review by gobbleturk — APPROVED** [review]

File: MaxText/configs/base.yml [hunk 6]
```diff
@@ -61,6 +61,13 @@ save_config_to_gcs: False
 dtype: "bfloat16"
 quantization: "" #defaults to no quantization, i.e. bf16. possible alternative setting is 'int8' or use fp8 to run with 8-bit floating-point GeMMs on NVIDIA GPUs.
 quantize_kvcache: False
+# Valid kv_quant_axis values:
+#   - "" is valid only when quantize_kvcache is False
+#   - "dkv" over kv_dimensions
```
Comment (yeandy) [inline MaxText/configs/base.yml:66]: Can we be be more descriptive on what `kv_dimensions` is?

Wondering if a separate writeup/README (with simple images or other points)outside of this code would be useful to explain the differences. Such as  on `dkv` vs `heads_and_dkv` perf/accuracy tradeoff.
